File: BFS-Course Schedule.cs
There are a total of n courses you have to take, labeled from 0 to n - 1.

Some courses may have prerequisites, 
for example to take course 0 you have to first take course 1, which is expressed as a pair: [0,1]

Given the total number of courses and a list of prerequisite pairs, is it possible for you to finish all courses?

For example:

2, [[1,0]]
There are a total of 2 courses to take. To take course 1 you should have finished course 0. So it is possible.

2, [[1,0],[0,1]]
There are a total of 2 courses to take. To take course 1 you should have finished course 0, and to take course 0 you should also have finished course 1. So it is impossible.

Note:
The input prerequisites is a graph represented by a list of edges, not adjacency matrices. Read more about how a graph is represented.


public class Solution {
    // check there is cycle is the directd graph -> BFS
    public bool CanFinish(int numCourses, int[,] prerequisites) {
        Dictionary<int, IList<int>> matrix = new Dictionary<int, IList<int>>(); // adjacency matrices to represent the graph
        int[] indegree = new int[numCourses];
        for(int i = 0; i < prerequisites.GetLength(0); i++){
            int current = prerequisites[i,0]; 
            int prerequisite = prerequisites[i,1];
            if(!matrix.ContainsKey(prerequisite)){
                matrix[prerequisite] = new List<int>();
            }
            if(!matrix[prerequisite].Contains(current)) indegree[current]++;  // duplicate edges in test
            matrix[prerequisite].Add(current);
        }
        
        Queue<int> queue = new Queue<int>();
        for(int i = 0; i < numCourses; i++){
            if(indegree[i] == 0){
                queue.Enqueue(i);
            }
        }
        
        int count = 0;
        while(queue.Count > 0){
            int prerequisite = queue.Dequeue();
            count++;
            if(matrix.ContainsKey(prerequisite)){
                foreach(int j in matrix[prerequisite]){
                    indegree[j]--;
                    if(indegree[j] == 0){
                        queue.Enqueue(j);
                    }
                }
            }
        }
        return count == numCourses;
    }
}

File: BFS-Course Schedule.java
There are a total of numCourses courses you have to take, labeled from 0 to numCourses - 1. 
    You are given an array prerequisites where prerequisites[i] = [ai, bi] indicates that you must take course bi first if you want to take course ai.

For example, the pair [0, 1], indicates that to take course 0 you have to first take course 1.
Return true if you can finish all courses. Otherwise, return false.

 
Example 1:
Input: numCourses = 2, prerequisites = [[1,0]]
Output: true
Explanation: There are a total of 2 courses to take. 
To take course 1 you should have finished course 0. So it is possible.
    
Example 2:
Input: numCourses = 2, prerequisites = [[1,0],[0,1]]
Output: false
Explanation: There are a total of 2 courses to take. 
To take course 1 you should have finished course 0, and to take course 0 you should also have finished course 1. So it is impossible.


class Solution {
    // 构建adjancency matrix和numPrereq array
    // numPrereq为0的表示可以take，没有前置条件；
    // BFS: queue中存储inumPrereq为0的元素；根据matrix依次访问并更新numPrereq
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        // build up data structure
        int n = prerequisites.length;
        Map<Integer, List<Integer>> matrix = new HashMap<>();
        int[] numPrereq = new int[numCourses];
        for(int i = 0; i < n; i++) {
            int cur = prerequisites[i][0];
            int prev = prerequisites[i][1];
            if (!matrix.containsKey(prev)) {
                matrix.put(prev, new ArrayList<>());
            }
            List<Integer> l = matrix.get(prev);
            l.add(cur);
            matrix.put(prev, l);
            numPrereq[cur]++;
        }
        Queue<Integer> queue = new LinkedList<>();
        for(int i = 0; i < numCourses; i++) {
            if (numPrereq[i] == 0) {
                queue.add(i);
            }
        }

        // BFS: take course & update indegree
        int count = 0;
        while(!queue.isEmpty()) {
            int cur = queue.poll();
            count++;
            if (matrix.containsKey(cur)) {
                List<Integer> l = matrix.get(cur);
                for (Integer i : l) {
                    numPrereq[i]--;
                    if (numPrereq[i] == 0) {
                        queue.add(i);
                    }
                }
            }
        }
        return numCourses == count;
    }
}
